1-Draft/RFCNNNN-Side-By-Side-Support.md

```diff
@@ -0,0 +1,253 @@
+
+    As a user,
+    I can use the command line to see which versions of PowerShell are installed,
+    so that I can know what versions are available to me to use.
+
+    As a user,
+    I can launch the version of PowerShell I need from the command line,
+    so that I can use automation built and tested against specific versions of PowerShell while still being able to use the latest version on the same system.
+
+## User Experience
+
+### Installing a specific version of PowerShell
+
+The end goal is for users to be easily able to do any of the following with a
+PowerShell installer:
+
+* install the latest version of PowerShell side-by-side with an existing
+version from a UI (e.g. `.msi`, `.pkg`, `.deb`)
+* install the latest version of PowerShell side-by-side with an existing
+version from the command line (e.g. `sudo apt-get install ...`, `brew cask
+install ...`, `choco install ...`)
+
+_Note: I'm a little uncertain of these details here, so please correct me where
+appropriate._
+
+The current release process seems to include updated published versions of
+packages in the various package repositories (for `apt-get`, homebrew, and
+Chocolatey). The problem with this approach is that it does not support
+side-by-side installations at the command line. For example, if I have
+PowerShell Core 6.2.0 installed on macOS and I want to install PowerShell Core
+6.2.1 side-by-side, I cannot. I can upgrade my installed version, reinstall
+the current version, or install a preview version (since that comes from a
+different package name), but I cannot install a different version side-by-side.
+
+Instead, if the release process not only updated the latest package but also
+created a version-specific package in the various repositories, then users
+could specify a specific version that they want to install (e.g. `brew cask
+install powershell@6.2.0`) to install different versions of PowerShell
+side-by-side with one another, with the versions automatically being installed
+in an appropriate version folder (which they are already anyway). Chocolatey
+already has support for installing a specific version, but it does not seem to
+work at the moment (I tried installing PowerShell Core 6.2.0 and ended up with
+6.2.1, with an error indicating that 6.2.0 was not installed).
+
+The UI installers should provide options to upgrade the latest installed
+version of PowerShell or install a new version side-by-side.
+
+### Listing installed versions of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, returns a list of installed
+# versions and their installation location, and then exits
+pwsh -ListVersions
+# This does the same thing
+pwsh -lv
+```
+
+```output
+6.2.1 [/usr/local/microsoft/powershell/6.2.1/]
+7.0.100-preview1 [/usr/local/microsoft/powershell/7.0.100-preview1]
+```
+
+```powershell
+
+# This returns a list of installed versions and their installation location as
```

Comment:
I think it should be explicitly stated that this will only look in the expected installed locations and not search the entire filesystem.  This means that custom deployments of the .zip/.tgz won't be found unless they are expanded in the same locations.

Reply:
Maybe use a file inventory that each version of PowerShell updates during install/uninstall. This method could read the file and list the versions there. Maybe with some validation that the source truly exists. Or detect Choco and list versions that way.

Reply:
if Windows is included here, the registry should be examined, IMO.